### ISSUE TITLE
[Discussion] How to rework `Publisher.withLatestFrom` to handle lone, upstream completions.

### ISSUE THREAD

**Issue by jasdev** [CombineCommunity/CombineExt#87]
👋🏽 I’m working on a post about implementing [`Publisher.withLatestFrom`](https://github.com/CombineCommunity/CombineExt/blob/730d2724b2a11e32652b1e35f5697d1d7677e17c/Sources/Operators/WithLatestFrom.swift) as a composed operator and — in making sure it matched Rx’s implementation — I noticed that Ext’s doesn’t finish when upstream completes (before the operator’s argument emits). Below are snippets with Rx’s and Ext’s current handling of this scenario:

#### `ObservableType.withLatestFrom` on a lone upstream completion.

```swift
import RxSwift

let disposeBag = DisposeBag()

let first = PublishSubject<Int>()
let second = PublishSubject<Int>()

first
	.withLatestFrom(second)
	.debug()
	.subscribe()
	.disposed(by: disposeBag)

first.onCompleted()

// Outputs (trimmed):
//
// ```
// -> subscribed
// -> Event completed
// -> isDisposed
// ``
```

#### `Publisher.withLatestFrom`’s handling of the same scenario.

```swift
final class WithLatestFromTests: XCTestCase {
	// …

	func testWithLatestFromCompletion() {
		let subject1 = PassthroughSubject<Int, Never>()
		let subject2 = PassthroughSubject<Int, Never>()
		var results = [Int]()
		var completed = false

		subscription = subject1
			.withLatestFrom(subject2)
			.sink(
				receiveCompletion: { _ in completed  = true },
				receiveValue: { results.append($0) }
			)

		subject1.send(completion: .finished)
		XCTAssertTrue(completed) // ❌ “`XCTAssertTrue` failed.”
		XCTAssertTrue(results.isEmpty)
	}

	// Similarly on analog tests for other arities.
}
```

Should we update the package to match Rx’s behavior? If so, I’ve got [a sketch](https://github.com/CombineCommunity/CombineExt/compare/main...jasdev:jasdev/composed-with-latest-from?expand=1&w=1) that passes an updated test suite with these scenarios over on a fork that I’d be happy to PR, but since the `WithLatestFrom` conformance is already in place, I wanted to check first to see if updating it in-place is easier. 🍻

**Comment by freak4pc** [comment]
Yup, the completion of `subject1` should terminate the entire dependent subscription. 
Sounds like a wrong "holding" by `WithLatestFrom`.

**Comment by jasdev** [comment]
10-4! Any tips on updating `WithLatestFrom` to handle this case? I’d be glad to help drive the implementation. ✅

**Comment by freak4pc** [comment]
@jasdev Hey there! Sorry for the huge delay, I'm in between jobs now so have a bit of time to look into this. 

The main issue is somewhere in between Sink and DemandBuffer. Basically the problem is this missing condition:
<img width="577" alt="image" src="https://user-images.githubusercontent.com/605076/120105289-f799b900-c160-11eb-86b0-ba6ced4c62ab.png">

But this breaks other publishers that are actually counting on the fact that a different publisher's completion shouldn't kill the entire publication. I need to look if all of these cases are actually valid and make a decision on how to fix this the best way (or perhaps create a sort of exclusion/inclusion for this case for specific publishers). 

Ill send more info soon, hopefully :)

**Comment by jasdev** [comment]
No problem and sounds good! If it’s easier, [the sketch](https://github.com/CombineCommunity/CombineExt/compare/main...jasdev:jasdev/composed-with-latest-from?expand=1&w=1) I linked to above can reduce the surface area of the change to help drive things forward. 👌🏽

**Comment by freak4pc** [comment]
@jasdev Sorry for the long delay here. I just don't have the time to dig into the details and think that if the test passes fine, let's pull in your variation of the composed operator into a PR and merge that in. Do you have a few moments to do that ? Thanks!

**Comment by freak4pc** [comment]
Merged in #105
